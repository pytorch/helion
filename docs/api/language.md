# Language Module

The `helion.language` module contains the core DSL constructs for writing GPU kernels.

## Loop Constructs

### tile()

```{eval-rst}
.. currentmodule:: helion.language

.. autofunction:: tile
```

The `tile()` function is the primary way to create parallel loops in Helion kernels. It provides several key features:

**Tiling Strategies**: The exact tiling strategy is determined by a Config object, typically created through autotuning. This allows for:
- Multidimensional tiling
- Index swizzling for cache locality
- Dimension reordering
- Flattening of iteration spaces

**Usage Patterns**:

```python
# Simple 1D tiling
for tile in hl.tile(1000):
    # tile.begin, tile.end, tile.block_size are available
    # Load entire tile (not just first element)
    data = tensor[tile]  # or hl.load(tensor, tile) for explicit loading
```

```python
# 2D tiling
for tile_i, tile_j in hl.tile([height, width]):
    # Each tile represents a portion of the 2D space
    pass
```

```python
# With explicit begin/end/block_size
for tile in hl.tile(0, 1000, block_size=64):
    pass
```

**Grid vs Loop Behavior**:
- When used at the top level of a kernel function, `tile()` becomes the grid of the kernel (parallel blocks)
- When used nested inside another loop, it becomes a sequential loop within each block

### grid()

```{eval-rst}
.. autofunction:: grid
```

The `grid()` function iterates over individual indices rather than tiles. It's equivalent to `tile(size, block_size=1)` but returns scalar indices instead of tile objects.

### static_range()

```{eval-rst}
.. autofunction:: static_range
```

`static_range()` behaves like a compile-time unrolled range for small loops. It hints the compiler to fully unroll the loop body where profitable.

## Memory Operations

### load()

```{eval-rst}
.. autofunction:: load
```

### store()

```{eval-rst}
.. autofunction:: store
```

### atomic_add()

```{eval-rst}
.. autofunction:: atomic_add
```

### atomic_and()

```{eval-rst}
.. autofunction:: atomic_and
```

### atomic_or()

```{eval-rst}
.. autofunction:: atomic_or
```

### atomic_xor()

```{eval-rst}
.. autofunction:: atomic_xor
```

### atomic_xchg()

```{eval-rst}
.. autofunction:: atomic_xchg
```

### atomic_max()

```{eval-rst}
.. autofunction:: atomic_max
```

### atomic_min()

```{eval-rst}
.. autofunction:: atomic_min
```

### atomic_cas()

```{eval-rst}
.. autofunction:: atomic_cas
```

## Inline Assembly

### inline_asm_elementwise()

```{eval-rst}
.. autofunction:: inline_asm_elementwise
```

Executes target-specific inline assembly on elements of one or more tensors with broadcasting and optional packed processing.

### inline_triton()

```{eval-rst}
.. autofunction:: inline_triton
```

Embeds small Triton code snippets directly inside a Helion kernel. Common indentation is removed automatically, placeholders are replaced using ``str.format`` with tuple or dict arguments, and the final line in the snippet becomes the return value. Provide tensors (or tuples of tensors) via ``output_like`` so Helion knows the type of the return value.

## Tensor Creation

### zeros()

```{eval-rst}
.. autofunction:: zeros
```

### full()

```{eval-rst}
.. autofunction:: full
```

### arange()

See {func}`~helion.language.arange` for details.

## Tunable Parameters

### register_block_size()

```{eval-rst}
.. autofunction:: register_block_size
```

### register_tunable()

```{eval-rst}
.. autofunction:: register_tunable
```

## Tile Operations

### Tile Class

```{eval-rst}
.. autoclass:: Tile
   :members:
   :undoc-members:
```

The `Tile` class represents a portion of an iteration space with the following key attributes:
- `begin`: Starting indices of the tile
- `end`: Ending indices of the tile
- `block_size`: Size of the tile in each dimension

## View Operations

### subscript()

```{eval-rst}
.. autofunction:: subscript
```

### split()

```{eval-rst}
.. autofunction:: split
```

### join()

```{eval-rst}
.. autofunction:: join
```

## StackTensor
### StackTensor class
```{eval-rst}
.. autoclass:: StackTensor
   :undoc-members:
```

### stacktensor_like
```{eval-rst}
.. autofunction:: stacktensor_like
```

## Reduction Operations

### reduce()

See {func}`~helion.language.reduce` for details.

## Scan Operations

### associative_scan()

See {func}`~helion.language.associative_scan` for details.

### cumsum()

See {func}`~helion.language.cumsum` for details.

### cumprod()

See {func}`~helion.language.cumprod` for details.

### tile_index()

```{eval-rst}
.. autofunction:: tile_index
```

### tile_begin()

```{eval-rst}
.. autofunction:: tile_begin
```

### tile_end()

```{eval-rst}
.. autofunction:: tile_end
```

### tile_block_size()

```{eval-rst}
.. autofunction:: tile_block_size
```

### tile_id()

```{eval-rst}
.. autofunction:: tile_id
```

## Synchronization


### signal()

```{eval-rst}
.. autofunction:: signal
```

### wait()

```{eval-rst}
.. autofunction:: wait
```

## Utilities

### device_print()

See {func}`~helion.language.device_print` for details.

## Constexpr Operations

### constexpr()

See {class}`~helion.language.constexpr` for details.

### specialize()

See {func}`~helion.language.specialize` for details.

## Matrix Operations

### dot()

See {func}`~helion.language.dot` for details.
