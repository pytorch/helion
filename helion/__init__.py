from __future__ import annotations

from triton import cdiv
from triton import next_power_of_2

from . import _compat as _compat_module  # noqa: F401  # side-effect import
from . import _logging
from . import exc
from . import language
from . import runtime
from .runtime import Config
from .runtime import Kernel
from .runtime import kernel
from .runtime import kernel as jit  # alias
from .runtime.settings import RefMode
from .runtime.settings import Settings

__all__ = [
    "Config",
    "Kernel",
    "RefMode",
    "Settings",
    "cdiv",
    "exc",
    "jit",
    "kernel",
    "language",
    "next_power_of_2",
    "runtime",
]

_logging.init_logs()
