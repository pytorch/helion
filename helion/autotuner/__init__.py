from __future__ import annotations

from .config_fragment import BooleanFragment as BooleanFragment
from .config_fragment import EnumFragment as EnumFragment
from .config_fragment import IntegerFragment as IntegerFragment
from .config_fragment import ListOf as ListOf
from .config_fragment import PowerOfTwoFragment as PowerOfTwoFragment
from .config_spec import ConfigSpec as ConfigSpec
from .differential_evolution import (
    DifferentialEvolutionSearch as DifferentialEvolutionSearch,
)
from .finite_search import FiniteSearch as FiniteSearch
from .local_cache import LocalAutotuneCache as LocalAutotuneCache
from .local_cache import StrictLocalAutotuneCache as StrictLocalAutotuneCache
from .pattern_search import PatternSearch as PatternSearch
from .random_search import RandomSearch as RandomSearch

search_algorithms = {
    "DifferentialEvolutionSearch": DifferentialEvolutionSearch,
    "FiniteSearch": FiniteSearch,
    "PatternSearch": PatternSearch,
    "RandomSearch": RandomSearch,
}
