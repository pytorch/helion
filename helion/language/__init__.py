from __future__ import annotations

from .constexpr import ConstExpr as constexpr  # noqa: F401
from .constexpr import specialize as specialize
from .creation_ops import arange as arange
from .creation_ops import full as full
from .creation_ops import zeros as zeros
from .device_print import device_print as device_print
from .inline_asm_ops import inline_asm_elementwise as inline_asm_elementwise
from .loops import grid as grid
from .loops import static_range as static_range
from .loops import tile as tile
from .matmul_ops import dot as dot
from .memory_ops import atomic_add as atomic_add
from .memory_ops import load as load
from .memory_ops import store as store
from .reduce_ops import reduce as reduce
from .scan_ops import associative_scan as associative_scan
from .scan_ops import cumprod as cumprod
from .scan_ops import cumsum as cumsum
from .signal_wait import signal as signal
from .signal_wait import wait as wait
from .slice_proxy import make_slice as make_slice
from .stack_tensor import StackTensor as StackTensor
from .stack_tensor import stacktensor_like as stacktensor_like
from .tile_ops import tile_begin as tile_begin
from .tile_ops import tile_block_size as tile_block_size
from .tile_ops import tile_end as tile_end
from .tile_ops import tile_id as tile_id
from .tile_ops import tile_index as tile_index
from .tile_proxy import Tile as Tile
from .tunable_ops import register_block_size as register_block_size
from .tunable_ops import register_reduction_dim as register_reduction_dim
from .tunable_ops import register_tunable as register_tunable
from .view_ops import subscript as subscript

_MEMORY_OPS = (store, load, atomic_add, wait, signal)
