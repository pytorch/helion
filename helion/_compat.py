from __future__ import annotations

import contextlib
import functools
from typing import Any
from typing import Callable
from typing import cast

import torch
from torch._inductor.runtime.hints import DeviceProperties
from torch._inductor.utils import triton_type
import triton
from triton.backends.compiler import BaseBackend
from triton.backends.compiler import GPUTarget
import triton.language as tl
import triton.runtime.jit as triton_jit

NativeSpecializeImpl = Callable[
    [type[BaseBackend], object, bool, bool, bool], tuple[object, ...]
]
CreateSpecializeImpl = Callable[
    [Callable[..., object]], Callable[..., tuple[object, ...]]
]


def _make_specialize_impl_wrapper(
    *,
    native_impl: NativeSpecializeImpl | None = None,
    create_factory: CreateSpecializeImpl | None = None,
) -> Callable[..., object]:
    if native_impl is None:
        native_impl = cast(
            "NativeSpecializeImpl | None",
            getattr(triton_jit, "native_specialize_impl", None),
        )
    if native_impl is None and create_factory is None:
        raise AttributeError("native_specialize_impl unavailable")

    def specialize_impl_wrapper(
        *args: object,
        **kwargs: object,
    ) -> object:
        specialize_extra = cast(
            "Callable[..., object] | None",
            kwargs.pop("specialize_extra", None),
        )
        kwargs.pop("specialize_zero_one", None)
        backend_param = kwargs.pop("backend", None)
        args_list: list[object] = list(args)
        backend_type: type[BaseBackend]
        if backend_param is None and args_list:
            first = args_list[0]
            if isinstance(first, type) and issubclass(first, BaseBackend):
                backend_type = first
                args_list.pop(0)
            elif isinstance(first, BaseBackend):
                backend_type = type(first)
                args_list.pop(0)
            else:
                backend_type = BaseBackend
        elif isinstance(backend_param, type) and issubclass(backend_param, BaseBackend):
            backend_type = backend_param
        elif isinstance(backend_param, BaseBackend):
            backend_type = type(backend_param)
        else:
            backend_type = BaseBackend

        arg = kwargs.pop("arg", None)
        if arg is None:
            if args_list:
                arg = args_list.pop(0)
            else:
                raise TypeError("specialize_impl() missing positional argument 'arg'")

        def _pop_flag(
            key: str,
            *,
            alt_keys: tuple[str, ...] = (),
            default: bool | None = None,
        ) -> bool:
            value = kwargs.pop(key, None)
            if value is None:
                for alt in alt_keys:
                    value = kwargs.pop(alt, None)
                    if value is not None:
                        break
            if value is None:
                if args_list:
                    value = args_list.pop(0)
                elif default is not None:
                    value = default
                else:
                    raise TypeError(f"specialize_impl() missing argument '{key}'")
            return bool(value)

        is_const = _pop_flag("is_const")
        specialize_value = _pop_flag(
            "specialize_value",
            alt_keys=("specialize",),
            default=True,
        )
        align = _pop_flag("align", default=True)

        if native_impl is not None:
            result = native_impl(
                backend_type,
                arg,
                is_const,
                specialize_value,
                align,
            )
            if specialize_extra is not None:
                with contextlib.suppress(Exception):
                    specialize_extra(arg)
        else:
            assert create_factory is not None

            def _call_specialize_extra(
                extra_arg: object,
                kind: object,
                *,
                align: bool = True,
            ) -> object:
                if specialize_extra is None:
                    return None
                try:
                    return specialize_extra(extra_arg)
                except TypeError:
                    try:
                        return specialize_extra(extra_arg, kind, align=align)
                    except Exception:
                        return None
                except Exception:
                    return None

            impl = create_factory(_call_specialize_extra)
            result = impl(
                arg,
                is_const=is_const,
                specialize_value=specialize_value,
                align=align,
            )
        return result

    return specialize_impl_wrapper


def _ensure_triton_specialize_impl_alias() -> None:
    if hasattr(triton_jit, "specialize_impl"):
        return
    if hasattr(triton_jit, "native_specialize_impl"):
        module: Any = triton_jit
        module.specialize_impl = _make_specialize_impl_wrapper()  # type: ignore[assignment]
        return
    if hasattr(triton_jit, "create_specialize_impl"):
        module: Any = triton_jit
        module.specialize_impl = _make_specialize_impl_wrapper(
            create_factory=triton_jit.create_specialize_impl,
        )  # type: ignore[assignment]


_ensure_triton_specialize_impl_alias()


def _ensure_backend_specialization_alias() -> None:
    if hasattr(BaseBackend, "get_arg_specialization"):
        return
    if hasattr(BaseBackend, "get_tensor_specialization"):
        BaseBackend.get_arg_specialization = BaseBackend.get_tensor_specialization  # type: ignore[attr-defined]


_ensure_backend_specialization_alias()


@functools.cache
def get_triton_find_paths_if() -> Callable[..., object]:
    if hasattr(triton_jit, "find_paths_if"):
        return triton_jit.find_paths_if
    if hasattr(triton_jit, "_find_paths_if"):
        return triton_jit._find_paths_if  # type: ignore[attr-defined]
    raise AttributeError("Unable to locate Triton find_paths_if helper")


@functools.cache
def get_triton_iterable_path() -> Callable[..., object]:
    if hasattr(triton_jit, "get_iterable_path"):
        return triton_jit.get_iterable_path
    if hasattr(triton_jit, "_get_iterable_path"):
        return triton_jit._get_iterable_path  # type: ignore[attr-defined]
    raise AttributeError("Unable to locate Triton get_iterable_path helper")


def supports_tensor_descriptor() -> bool:
    # call private func we can patch in testing
    return _supports_tensor_descriptor()


@functools.cache
def _supports_tensor_descriptor() -> bool:
    def _cuda_tensor_desc_available() -> bool:
        if not torch.cuda.is_available():
            return False
        major, _ = torch.cuda.get_device_capability(torch.cuda.current_device())
        return major >= 9

    def _xpu_tensor_desc_available() -> bool:
        if not torch.xpu.is_available():
            return False

        from packaging import version

        return version.parse(triton.__version__) >= version.parse("3.5")

    if not (_cuda_tensor_desc_available() or _xpu_tensor_desc_available()):
        return False

    return hasattr(triton.language, "make_tensor_descriptor") or hasattr(
        triton.language, "_experimental_make_tensor_descriptor"
    )


@functools.cache
def get_tensor_descriptor_fn_name() -> str:
    if hasattr(triton.language, "make_tensor_descriptor"):
        return "tl.make_tensor_descriptor"
    assert hasattr(triton.language, "_experimental_make_tensor_descriptor")
    return "tl._experimental_make_tensor_descriptor"


@functools.cache
def torch_dtype_to_tl(torch_dtype: torch.dtype) -> object:
    """Return the `triton.language` dtype that matches a `torch.dtype`."""
    name_str = triton_type(torch_dtype).replace("tl.", "")
    return getattr(tl, name_str)


def min_dot_size(
    device: torch.device, lhs: torch.dtype, rhs: torch.dtype
) -> tuple[int, int, int]:
    # call private func we can patch in testing
    return _min_dot_size(device, lhs, rhs)


@functools.cache
def _min_dot_size(
    device: torch.device, lhs: torch.dtype, rhs: torch.dtype
) -> tuple[int, int, int]:
    if device.type not in ["cuda", "xpu"]:
        # TODO(jansel): support other hardware backends properly besides CUDA and XPU
        return (16, 16, 16)

    if torch.xpu.is_available():
        from triton.backends.intel.compiler import (  # pyright: ignore[reportMissingImports]
            min_dot_size as min_dot_size_xpu,
        )

        device_properties = torch.xpu.get_device_properties()
        gpu_target_info = {
            k: getattr(device_properties, k)
            for k in device_properties.__dir__()
            if not k.startswith("_")
        }

        dot_size_val = min_dot_size_xpu(gpu_target_info)(
            torch_dtype_to_tl(lhs), torch_dtype_to_tl(rhs)
        )
        return tuple(int(v) for v in dot_size_val)  # pyright: ignore[reportReturnType]

    from triton.backends.nvidia.compiler import min_dot_size as min_dot_size_cuda

    props = DeviceProperties.create(device)
    return min_dot_size_cuda(
        GPUTarget(
            backend=props.type,
            arch=props.cc,
            warp_size=props.warp_size or 32,
        )
    )(torch_dtype_to_tl(lhs), torch_dtype_to_tl(rhs))


def warps_to_threads(num_warps: int) -> int:
    if torch.cuda.is_available():
        props = DeviceProperties.create(
            torch.device("cuda", torch.cuda.current_device())
        )
        return num_warps * (props.warp_size or 32)
    return num_warps * 32
