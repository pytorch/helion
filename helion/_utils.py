from __future__ import annotations

import collections
from typing import Sequence

counters: collections.defaultdict[str, collections.Counter[str]] = (
    collections.defaultdict(collections.Counter)
)


def create_shape_matching_slices(
    shape1: Sequence[int], shape2: Sequence[int]
) -> tuple[slice, ...]:
    """Create slices to match the smaller of two shapes.

    This is used for masking tensors to compatible shapes by taking the
    minimum size in each dimension.

    Args:
        shape1: First shape (can be torch.Size or any sequence of ints)
        shape2: Second shape (can be torch.Size or any sequence of ints)

    Returns:
        Tuple of slices that can be used to index a tensor
    """
    return tuple(slice(0, min(d1, d2)) for d1, d2 in zip(shape1, shape2, strict=False))


def convert_size_arg(size: object) -> object:
    """Convert a size argument that may contain RefTile objects.

    Handles:
    - Single RefTile -> int (block_size)
    - List/tuple containing RefTiles -> list with converted sizes
    - Other values -> unchanged
    """
    # Import here to avoid circular dependency
    from helion.language.ref_tile import RefTile

    if isinstance(size, (list, tuple)):
        return [convert_size_arg(item) for item in size]
    if isinstance(size, RefTile):
        return size._block_size
    return size


def convert_tile_indices_to_slices(index: object) -> object:
    """Convert RefTile objects in index to their corresponding slice objects.

    Args:
        index: Index that may contain RefTile objects or tuples of indices

    Returns:
        Index with RefTile objects replaced by their slice objects
    """
    # Import here to avoid circular dependency
    from helion.language.ref_tile import RefTile

    def _extract_slice(obj: object) -> object:
        return obj._slice if isinstance(obj, RefTile) else obj

    if isinstance(index, tuple):
        return tuple(_extract_slice(idx) for idx in index)
    return _extract_slice(index)
