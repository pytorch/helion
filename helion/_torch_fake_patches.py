from __future__ import annotations

from typing import Iterable, Sequence

import sympy
import torch
import triton._utils as triton_utils
import triton.language as triton_lang
import triton.language.core as triton_core
from sympy.core.relational import Relational
from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
import torch._subclasses.fake_tensor as fake_tensor_module
from torch._inductor import sizevars as inductor_sizevars
from torch._inductor.utils import sympy_subs
from torch._subclasses import fake_impls
import torch.fx.experimental.symbolic_shapes as symbolic_shapes
from torch.fx.experimental.symbolic_shapes import guard_or_false
from torch.fx.experimental.symbolic_shapes import SymBool

_SUM_DIM = torch.ops.aten.sum.dim_IntList

_original_dispatch_impl = FakeTensorMode._dispatch_impl


def _normalize_dims(dim_arg: object, rank: int) -> tuple[int, ...] | None:
    if dim_arg is None:
        return tuple(range(rank))
    if isinstance(dim_arg, (list, tuple)):
        dims = tuple(int(d) for d in dim_arg)
    else:
        dims = (int(dim_arg),)
    normalized = []
    for dim in dims:
        if dim < 0:
            dim += rank
        if dim < 0 or dim >= rank:
            return None
        normalized.append(dim)
    return tuple(sorted(set(normalized)))


def _canonicalize_sum_output(
    fake_mode: FakeTensorMode,
    result: object,
    args: Sequence[object],
    kwargs: dict[str, object],
) -> object:
    if not isinstance(result, FakeTensor):
        return result
    if not args:
        return result

    input_tensor = args[0]
    if not isinstance(input_tensor, FakeTensor):
        return result

    dim_arg = None
    if len(args) > 1:
        dim_arg = args[1]
    elif "dim" in kwargs:
        dim_arg = kwargs["dim"]

    keepdim = bool(kwargs.get("keepdim", False))

    dims = _normalize_dims(dim_arg, input_tensor.dim())
    if dims is None:
        return result

    input_sizes: list[object] = list(input_tensor.size())
    output_sizes: list[object] = list(result.size())

    def _canonical_size(dim: object) -> object:
        return dim

    if keepdim:
        if len(output_sizes) != len(input_sizes):
            return result
        new_sizes = [output_sizes[i] for i in range(len(output_sizes))]
        for idx in range(len(new_sizes)):
            if idx not in dims:
                new_sizes[idx] = _canonical_size(input_sizes[idx])
    else:
        remaining = [i for i in range(len(input_sizes)) if i not in dims]
        if len(remaining) != len(output_sizes):
            return result
        new_sizes = [_canonical_size(input_sizes[i]) for i in remaining]

    try:
        meta = torch.empty_strided(
            new_sizes,
            result.stride(),
            dtype=result.dtype,
            device=torch.device("meta"),
        )
    except Exception:
        return result

    meta.requires_grad_(result.requires_grad)

    replacement = FakeTensor(
        fake_mode,
        meta,
        result.fake_device,
        constant=result.constant,
        real_tensor=result.real_tensor,
        pytype=result.pytype,
        dispatch_keys=result.dispatch_keys,
    )
    return replacement


def _patched_dispatch_impl(
    self: FakeTensorMode,
    func: torch._ops.OpOverload,
    types: Sequence[type],
    args: Sequence[object],
    kwargs: dict[str, object],
) -> object:
    if func is torch.ops.aten.add.Tensor and len(args) >= 2:
        lhs, rhs = args[0], args[1]
        if isinstance(lhs, FakeTensor) and isinstance(rhs, FakeTensor):
            lhs_sizes = tuple(lhs.size())
            rhs_sizes = tuple(rhs.size())
            if len(lhs_sizes) == len(rhs_sizes):
                try:
                    from helion._compiler.compile_environment import CompileEnvironment

                    env = CompileEnvironment.current()
                except Exception:
                    env = None

                def _resolve(dim: object) -> object:
                    if env is None or not isinstance(dim, torch.SymInt):
                        return dim
                    expr = env.shape_env.simplify(dim._sympy_())
                    seen: set[sympy.Expr] = set()
                    while expr in env.specialized_aliases and expr not in seen:
                        seen.add(expr)
                        expr = env.specialized_aliases[expr]
                    return expr

                if all(
                    (ldim == rdim)
                    or (
                        isinstance(ldim, torch.SymInt)
                        and isinstance(rdim, torch.SymInt)
                        and _resolve(ldim) == _resolve(rdim)
                    )
                    for ldim, rdim in zip(lhs_sizes, rhs_sizes)
                ):
                    try:
                        meta = torch.empty_strided(
                            lhs_sizes,
                            rhs.stride(),
                            dtype=rhs.dtype,
                            device=torch.device("meta"),
                        )
                    except Exception:
                        meta = None
                    if meta is not None:
                        meta.requires_grad_(rhs.requires_grad)
                        rhs = FakeTensor(
                            self,
                            meta,
                            rhs.fake_device,
                            constant=rhs.constant,
                            real_tensor=rhs.real_tensor,
                            pytype=rhs.pytype,
                            dispatch_keys=rhs.dispatch_keys,
                        )
                        args = (lhs, rhs, *args[2:])
    result = _original_dispatch_impl(self, func, types, args, kwargs)
    if func is _SUM_DIM:
        result = _canonicalize_sum_output(self, result, args, dict(kwargs))
    return result


FakeTensorMode._dispatch_impl = _patched_dispatch_impl  # type: ignore[assignment]


def _helion_infer_size(a: Sequence[object], b: Sequence[object]) -> tuple[object, ...]:
    dims_a = len(a)
    dims_b = len(b)
    ndim = max(dims_a, dims_b)
    expanded = [0] * ndim

    def _symbols_match(x: object, y: object) -> bool:
        if not isinstance(x, torch.SymInt) or not isinstance(y, torch.SymInt):
            return False
        node_x = getattr(x, "node", None)
        node_y = getattr(y, "node", None)
        shape_env = getattr(node_x, "shape_env", None)
        if shape_env is None or getattr(node_y, "shape_env", None) is not shape_env:
            return False
        expr_x = node_x.expr
        expr_y = node_y.expr
        try:
            evaluated = shape_env._maybe_evaluate_static(sympy.Eq(expr_x, expr_y))
            if evaluated is not None:
                return bool(evaluated)
        except Exception:
            pass

        val_x = shape_env.var_to_val.get(expr_x)
        val_y = shape_env.var_to_val.get(expr_y)
        if val_x is not None and val_y is not None and val_x == val_y:
            return True

        try:
            from helion._compiler.compile_environment import CompileEnvironment

            env = CompileEnvironment.current()
        except Exception:
            env = None
        if env is not None:
            spec_vals = env.specialized_values
            vx = spec_vals.get(expr_x)
            vy = spec_vals.get(expr_y)
            if vx is not None and vy is not None and vx == vy:
                return True

        return False

    for i in range(ndim - 1, -1, -1):
        offset = ndim - 1 - i
        dim_a = dims_a - 1 - offset
        dim_b = dims_b - 1 - offset
        size_a = a[dim_a] if dim_a >= 0 else 1
        size_b = b[dim_b] if dim_b >= 0 else 1

        broadcastable = (
            guard_or_false(size_a == 1)
            or guard_or_false(size_b == 1)
            or size_a == size_b
            or _symbols_match(size_a, size_b)
        )

        torch._check(
            broadcastable,
            lambda size_a=size_a, size_b=size_b, i=i: (
                f"The size of tensor a ({size_a}) must match the size of tensor b ({size_b}) "
                f"at non-singleton dimension {i})"
            ),
        )

        expanded[i] = size_b if guard_or_false(size_a == 1) else size_a

    return tuple(expanded)


fake_impls.infer_size = _helion_infer_size

_orig_get_fast_op_impls = fake_impls.get_fast_op_impls


def _helion_get_fast_op_impls() -> dict[object, object]:
    impls = _orig_get_fast_op_impls()
    impls.pop(torch.ops.aten.add.Tensor, None)
    return impls


fake_impls.get_fast_op_impls = _helion_get_fast_op_impls
fake_tensor_module.get_fast_op_impls = _helion_get_fast_op_impls


def _helion_expr_equal(lhs: sympy.Expr, rhs: sympy.Expr) -> bool:
    try:
        from helion._compiler.compile_environment import CompileEnvironment

        env = CompileEnvironment.current()
    except Exception:
        return False

    lhs = env.resolve_alias(lhs)
    rhs = env.resolve_alias(rhs)

    values = {k: sympy.Integer(v) for k, v in env.specialized_values.items()}
    if not values:
        return False
    try:
        lhs_val = sympy_subs(lhs, values)
        rhs_val = sympy_subs(rhs, values)
    except Exception:
        return False
    print("helion inductor eq", lhs, rhs, lhs_val, rhs_val)
    try:
        lhs_val = sympy.Integer(lhs_val)
        rhs_val = sympy.Integer(rhs_val)
    except Exception:
        pass
    return lhs_val == rhs_val


_orig_expect_true = inductor_sizevars.SizeVarAllocator.expect_true


def _helion_expect_true(self: inductor_sizevars.SizeVarAllocator, expr: sympy.Expr) -> bool:
    if isinstance(expr, sympy.Equality):
        if _helion_expr_equal(expr.lhs, expr.rhs):
            return True
    return _orig_expect_true(self, expr)


inductor_sizevars.SizeVarAllocator.expect_true = _helion_expect_true

_orig_guard_or_false = symbolic_shapes.guard_or_false


def _helion_guard_or_false(a: object) -> bool:
    if isinstance(a, SymBool):
        expr = getattr(a.node, "expr", None)
        if isinstance(expr, Relational):
            if isinstance(expr, sympy.Equality):
                if _helion_expr_equal(expr.lhs, expr.rhs):
                    return True
            elif getattr(expr, "rel_op", None) == "!=":
                if _helion_expr_equal(expr.lhs, expr.rhs):
                    return False
    return _orig_guard_or_false(a)


symbolic_shapes.guard_or_false = _helion_guard_or_false  # type: ignore[assignment]
guard_or_false = symbolic_shapes.guard_or_false  # keep module-level alias in sync

_orig_validate_block_shape = triton_utils.validate_block_shape


def _helion_validate_block_shape(shape: list[int]) -> int:
    numel = 1
    for i, d in enumerate(shape):
        if not isinstance(d, int):
            raise TypeError(
                f"Shape element {i} must have type `constexpr[int]`, got `constexpr[{type(d)}]`"
            )
        numel *= d
    if numel > triton_utils.TRITON_MAX_TENSOR_NUMEL:
        raise ValueError(
            f"numel ({numel}) exceeds triton maximum tensor numel "
            f"({triton_utils.TRITON_MAX_TENSOR_NUMEL})"
        )
    return numel


triton_utils.validate_block_shape = _helion_validate_block_shape  # type: ignore[assignment]
triton_core.validate_block_shape = _helion_validate_block_shape  # type: ignore[assignment]
