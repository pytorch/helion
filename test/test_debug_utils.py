from __future__ import annotations

import linecache
import os
import unittest

import pytest
import torch

import helion
from helion._testing import DEVICE
from helion._testing import RefEagerTestDisabled
from helion._testing import TestCase
import helion.language as hl


@pytest.fixture(autouse=True)
def _store_capfd_on_class(request, capfd):
    """
    Expose pytest's capfd fixture as `self._capfd` inside the TestDebugUtils class
    (works for unittest.TestCase-style tests).
    """
    if request.cls is not None:
        request.cls._capfd = capfd


class TestDebugUtils(RefEagerTestDisabled, TestCase):
    def test_print_repro_env_var(self):
        """Ensure HELION_PRINT_REPRO=1 emits an executable repro script."""
        original = os.environ.get("HELION_PRINT_REPRO")
        os.environ["HELION_PRINT_REPRO"] = "1"
        try:

            @helion.kernel(
                config=helion.Config(
                    block_sizes=[2, 2],
                    flatten_loops=[False],
                    indexing=["pointer", "pointer"],
                    l2_groupings=[1],
                    load_eviction_policies=[""],
                    loop_orders=[[0, 1]],
                    num_stages=1,
                    num_warps=4,
                    pid_type="flat",
                    range_flattens=[None],
                    range_multi_buffers=[None],
                    range_num_stages=[0],
                    range_unroll_factors=[0],
                ),
                static_shapes=True,
            )
            def kernel1(x: torch.Tensor) -> torch.Tensor:
                out = torch.empty_like(x)
                m, n = x.shape
                for tile_m, tile_n in hl.tile([m, n]):
                    out[tile_m, tile_n] = x[tile_m, tile_n] + 1
                return out

            torch.manual_seed(0)
            x = torch.randn([2, 2], dtype=torch.float32, device=DEVICE)

            if hasattr(self, "_capfd"):
                self._capfd.readouterr()

            result = kernel1(x)
            torch.testing.assert_close(result, x + 1)

            if not hasattr(self, "_capfd"):
                return  # Cannot test without capture

            captured = "".join(self._capfd.readouterr())

            # Extract repro script
            lines = captured.splitlines()
            start = next(
                i
                for i, line in enumerate(lines)
                if "# === HELION KERNEL REPRO ===" in line
            )
            end = next(
                i
                for i, line in enumerate(lines[start:], start)
                if "# === END HELION KERNEL REPRO ===" in line
            )
            repro_script = "\n".join(lines[start : end + 1])

            # Normalize range_warp_specializes=[None] to [] for comparison
            normalized_script = repro_script.replace(
                "range_warp_specializes=[None]", "range_warp_specializes=[]"
            )

            # Verify repro script matches expected script
            self.assertExpectedJournal(normalized_script)

            # Extract the actual code (without the comment markers) for execution
            repro_lines = repro_script.splitlines()
            code_start = 1 if repro_lines[0].startswith("# === HELION") else 0
            code_end = len(repro_lines) - (
                1 if repro_lines[-1].startswith("# === END") else 0
            )
            repro_code = "\n".join(repro_lines[code_start:code_end])

            # Setup linecache so inspect.getsource() works on exec'd code
            filename = "<helion_repro_test>"
            linecache.cache[filename] = (
                len(repro_code),
                None,
                [f"{line}\n" for line in repro_code.splitlines()],
                filename,
            )

            # Execute the repro script
            namespace = {}
            exec(compile(repro_code, filename, "exec"), namespace)

            # Call the generated helper and verify it runs successfully
            helper = namespace["helion_repro_caller"]
            repro_result = helper()

            # Verify the output
            torch.testing.assert_close(repro_result, x + 1)

            linecache.cache.pop(filename, None)
        finally:
            if original is None:
                os.environ.pop("HELION_PRINT_REPRO", None)
            else:
                os.environ["HELION_PRINT_REPRO"] = original


if __name__ == "__main__":
    unittest.main()
