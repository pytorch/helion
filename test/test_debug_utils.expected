This file is automatically generated by assertExpectedJournal calls in test_debug_utils.py.
Update expected outputs by running tests with the EXPECTTEST_ACCEPT=1 environment variable set.

--- assertExpectedJournal(TestDebugUtils.test_print_repro_env_var)
# === HELION KERNEL REPRO ===
import helion
import helion.language as hl
import torch
from torch._dynamo.testing import rand_strided

@helion.kernel(config=helion.Config(block_sizes=[2, 2], flatten_loops=[False], indexing=['pointer', 'pointer'], l2_groupings=[1], load_eviction_policies=[''], loop_orders=[[0, 1]], num_stages=1, num_warps=4, pid_type='flat', range_flattens=[None], range_multi_buffers=[None], range_num_stages=[0], range_unroll_factors=[0], range_warp_specializes=[]), static_shapes=True)
def kernel1(x: torch.Tensor) -> torch.Tensor:
    out = torch.empty_like(x)
    m, n = x.shape
    for tile_m, tile_n in hl.tile([m, n]):
        out[tile_m, tile_n] = x[tile_m, tile_n] + 1
    return out

def helion_repro_caller():
    torch.manual_seed(0)
    x = rand_strided((2, 2), (2, 1), dtype=torch.float32, device=DEVICE)
    return kernel1(x)
# === END HELION KERNEL REPRO ===
